File: packages/py/discover.py
"""discover.py — load the contract, connect over HTTP/2, pull the tool list.

    python discover.py
    GRPC_TARGET=host:port python discover.py
"""
import json
import os

import grpc

from proto import tools_pb2, tools_pb2_grpc


def fetch_tools() -> None:
    target = os.environ.get("GRPC_TARGET", "localhost:50051")
    print(f"Connecting to gRPC server at {target}...")

    # Insecure for local dev; use grpc.secure_channel() in production.
    with grpc.insecure_channel(target) as channel:
        stub = tools_pb2_grpc.DiscoveryServiceStub(channel)
        try:
            response = stub.GetAvailableTools(tools_pb2.EmptyRequest())
        except grpc.RpcError as err:
            print(f"❌ Failed to discover tools: {err.details()}")
            raise SystemExit(1)

    tools = [
        {
            "name": t.name,
            "description": t.description,
            "input_schema": t.input_schema,
        }
        for t in response.tools
    ]
    print("✅ Discovered Tools:")
    print(json.dumps(tools, indent=2))


if __name__ == "__main__":
    fetch_tools()

File: packages/ts/src/discover.ts
// discover.ts — dynamically load the contract, connect over HTTP/2, and pull
// the tool list. The callback-style unary call is wrapped in a Promise.
//
//   npm run discover
//   GRPC_TARGET=host:port npm run discover
import * as grpc from '@grpc/grpc-js';
import { DiscoveryService, type ToolListResponse } from './proto.js';

function fetchTools(): Promise<ToolListResponse> {
  const target = process.env.GRPC_TARGET ?? 'localhost:50051';

  // Insecure for local dev; use grpc.credentials.createSsl() in production.
  const client = new DiscoveryService(target, grpc.credentials.createInsecure());

  console.log(`Connecting to gRPC server at ${target}...`);

  return new Promise<ToolListResponse>((resolve, reject) => {
    client.GetAvailableTools(
      {},
      (error: grpc.ServiceError | null, response: ToolListResponse) => {
        client.close();
        if (error) {
          reject(error);
          return;
        }
        resolve(response);
      },
    );
  });
}

try {
  const response = await fetchTools();
  console.log('✅ Discovered Tools:');
  console.log(JSON.stringify(response.tools, null, 2));
} catch (error) {
  const err = error as grpc.ServiceError;
  console.error('❌ Failed to discover tools:', err.details || err.message);
  process.exitCode = 1;
}
